File: zetech_hostel_app/lib/models/room_model.dart
class Room {
  final String? id;
  final String? roomNumber;
  final String? hostel;
  final String? campus;
  final String? roomType;
  final int? capacity;
  final double? price;
  final String? status;
  final int? occupiedBeds;

  Room({this.id, this.roomNumber, this.hostel, this.campus, this.roomType, this.capacity, this.price, this.status, this.occupiedBeds});

  factory Room.fromJson(Map<String, dynamic> json) => Room(
    id: (json['_id'] ?? json['id'])?.toString(),
    roomNumber: (json['roomNumber'] ?? json['room_number'])?.toString(),
    hostel: json['hostel']?.toString(),
    campus: json['campus']?.toString(),
    roomType: (json['roomType'] ?? json['room_type'])?.toString(),
    capacity: json['capacity'] is int ? json['capacity'] : int.tryParse(json['capacity']?.toString() ?? ''),
    price: (json['price'] as num?)?.toDouble(),
    status: json['status']?.toString(),
    // Server returns the field as 'occupied' (count of approved bookings)
    occupiedBeds: json['occupied'] ?? json['occupiedBeds'] ?? json['occupied_beds'] ?? 0,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'roomNumber': roomNumber,
    'hostel': hostel,
    'campus': campus,
    'roomType': roomType,
    'capacity': capacity,
    'price': price,
    'status': status,
    'occupiedBeds': occupiedBeds,
  };
}

File: server/models/roomModel.js
const pool = require('../config/db');

const getAllRooms = () => {
    return pool.query(
        `SELECT r.*, 
         (SELECT COUNT(*) FROM bookings b WHERE b.room_id = r.id AND b.status = 'approved') AS occupied
         FROM rooms r ORDER BY r.hostel, r.room_number`
    );
};

const getRoomById = (id) => {
    return pool.query('SELECT * FROM rooms WHERE id = ?', [id]);
};

const getRoomBeds = (roomId) => {
    return pool.query('SELECT * FROM beds WHERE room_id = ? ORDER BY bed_number', [roomId]);
};

const createRoom = (room_number, hostel, campus, room_type, capacity, price) => {
    return pool.query(
        'INSERT INTO rooms (room_number, hostel, campus, room_type, capacity, price) VALUES (?, ?, ?, ?, ?, ?)',
        [room_number, hostel, campus, room_type || 'quad', capacity || 4, price]
    );
};

const updateRoom = (id, room_number, hostel, campus, room_type, capacity, price, status) => {
    return pool.query(
        'UPDATE rooms SET room_number=?, hostel=?, campus=?, room_type=?, capacity=?, price=?, status=? WHERE id=?',
        [room_number, hostel, campus, room_type, capacity, price, status, id]
    );
};

const deleteRoom = (id) => {
    return pool.query('DELETE FROM rooms WHERE id = ?', [id]);
};

const checkActiveBookings = (roomId) => {
    return pool.query(
        "SELECT COUNT(*) AS count FROM bookings WHERE room_id = ? AND status IN ('pending','approved')",
        [roomId]
    );
};

const getBedById = (bedId, roomId) => {
    return pool.query('SELECT * FROM beds WHERE id = ? AND room_id = ? AND status = ?', [bedId, roomId, 'available']);
};

const updateBedStatus = (bedId, status) => {
    return pool.query('UPDATE beds SET status = ? WHERE id = ?', [status, bedId]);
};

module.exports = { getAllRooms, getRoomById, getRoomBeds, createRoom, updateRoom, deleteRoom, checkActiveBookings, getBedById, updateBedStatus };
